### ISSUE TITLE
gcc5 & gcc6 when MacOS::CLT not installed

### ISSUE THREAD

**Issue by ilg-ul** [Homebrew/homebrew-versions#1456]
I have a custom homebrew instance where I keep gcc5 and gcc6 for my test builds on macOS.

The homebrew instances are installed with several scripts, available from [gist](https://gist.github.com/ilg-ul/46407a070844f764dec6f27bde385797).

Recently I upgraded my mac to Sierra, and ran the `install-homebrew-gcc.sh` script, which completed properly.

But, when I tried to use the compilers, my build failed, claiming some system headers were not available.

After inspecting the formulas, I discovered this:

```
      unless MacOS::CLT.installed?
        # For Xcode-only systems, we need to tell the sysroot path.
        # "native-system-headers" will be appended
        args << "--with-native-system-header-dir=/usr/include"
        args << "--with-sysroot=#{MacOS.sdk_path}"
      end
```

In other words, if the Xcode Command Line Tools are not installed, the gcc build will not add the system headers. That was also my case, after the upgrade to Sierra, I installed Xcode, but forgot to install the new Command Line Tools.

The question is what would be the usability of the resulted incomplete compilers, since simple builds that include system headers will fail?

Suggestion:

- make the formula depend on the presence of MacOS::CLT; if CLT is not installed, abort the build; for special cases, possibly add a --force, to build an incomplete compiler.

**Comment by ilovezfs** [comment]
>The question is what would be the usability of the resulted incomplete compilers, since simple builds that include system headers will fail?

The headers are in the Xcode SDK folder:
```
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk/usr/include
```

**Comment by MikeMcQuaid** [comment]
> make the formula depend on the presence of MacOS::CLT

There's generally no reason to depend on the CLT because everything is available in Xcode itself too. `brew doctor` will warn you about the non-generally cases.

**Comment by ilg-ul** [comment]
sorry, I don't understand your answers, and I'm afraid you did not understand the problem.

if you install homebrew and gcc on a system without CLT, you do not get any warnings or errors, and after half an hour of waiting to compile the sources, you get a completely useless compiler, that will fail complaining about missing system header files.

is this the expected/desired behaviour?

in my opinion this is definitely time lost, because after you install CLT you need to reinstall (rebuild) gcc again in homebrew, to have configure add the missing option.

**Comment by ilovezfs** [comment]
>you do not get any warnings or errors

You get a warning:
```
Warning: Building gcc from source:
  The bottle needs the Xcode CLT to be installed.
```

@ilg-ul this is on a system without the CLT:

```
Josephs-Mac:~ joe$ brew test -vd gcc
/usr/local/Homebrew/Library/Homebrew/brew.rb (Formulary::FormulaLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gcc.rb
Testing gcc
==> Using the sandbox
/usr/bin/sandbox-exec -f /tmp/homebrew20161115-93679-1utyse.sb /System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gcc.rb -vd
/usr/local/Homebrew/Library/Homebrew/test.rb (Formulary::FromPathLoader): loading /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/gcc.rb
==> /usr/local/Cellar/gcc/6.2.0/bin/gcc-6 -o hello-c hello-c.c
==> /usr/local/Cellar/gcc/6.2.0/bin/g++-6 -o hello-cc hello-cc.cc
==> /usr/local/Cellar/gcc/6.2.0/bin/gfortran -c in.f90
==> /usr/local/Cellar/gcc/6.2.0/bin/gfortran -o test in.o
Josephs-Mac:~ joe$
```

The tests pass.

>if you install homebrew and gcc on a system without CLT, you do not get any warnings or errors, and after half an hour of waiting to compile the sources, you get a completely useless compiler, that will fail complaining about missing system header files.

What do you mean by "completely useless compiler, that will fail complaining about missing system header files"? Can you provide an example?

**Comment by ilg-ul** [comment]
> What do you mean by "completely useless compiler, that will fail complaining about missing system header files"? Can you provide an example?

my test is in C++ and uses `<cassert>`, which refers to `<assert.h>`, reported as missing by gcc-5 (and possibly 6, but I already deleted the bad install, so I cannot test).

so I suggest you try to include `<assert.h>`; if you built gcc with `MacOS::CLT.installed == false`, the `--with-native-system-header-dir=/usr/include --with-sysroot=#{MacOS.sdk_path}` were not defined, and the compile will fail.

**Comment by ilg-ul** [comment]
> this is on a system without the CLT:

@ilovezfs, could you run `gcc -v` and check if `--with-native-system-header-dir=/usr/include --with-sysroot=#{MacOS.sdk_path}` were present when you built the compiler?

**Comment by ilovezfs** [comment]
@ilg-ul 
```
Josephs-Mac:~ joe$ gcc-6 -v
Using built-in specs.
COLLECT_GCC=gcc-6
COLLECT_LTO_WRAPPER=/usr/local/Cellar/gcc/6.2.0/libexec/gcc/x86_64-apple-darwin16.1.0/6.2.0/lto-wrapper
Target: x86_64-apple-darwin16.1.0
Configured with: ../configure --build=x86_64-apple-darwin16.1.0 --prefix=/usr/local/Cellar/gcc/6.2.0 --libdir=/usr/local/Cellar/gcc/6.2.0/lib/gcc/6 --enable-languages=c,c++,objc,obj-c++,fortran --program-suffix=-6 --with-gmp=/usr/local/opt/gmp --with-mpfr=/usr/local/opt/mpfr --with-mpc=/usr/local/opt/libmpc --with-isl=/usr/local/opt/isl --with-system-zlib --enable-libstdcxx-time=yes --enable-stage1-checking --enable-checking=release --enable-lto --with-build-config=bootstrap-debug --disable-werror --with-pkgversion='Homebrew gcc 6.2.0' --with-bugurl=https://github.com/Homebrew/homebrew/issues --enable-plugin --disable-nls --enable-multilib --with-native-system-header-dir=/usr/include --with-sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
Thread model: posix
gcc version 6.2.0 (Homebrew gcc 6.2.0) 
Josephs-Mac:~ joe$
```

**Comment by ilg-ul** [comment]
@ilovezfs,

as I guessed, your compiler was built with `MacOS::CLT.installed`.

the problem I am reporting occurs when you do it without `MacOS::CLT.installed`.

to be sure we are speaking about the same formula, I suggest you try it with `gcc-5`.

**Comment by ilovezfs** [comment]
No, it's built with CLT uninstalled.

See the
```
 --with-native-system-header-dir=/usr/include --with-sysroot=/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX10.12.sdk
```
which is only passed when CLT is uninstalled.

**Comment by ilg-ul** [comment]
> which is only passed when CLT is uninstalled.

ah, yes, sorry, I understood it wrong.

anyway, I regret that I removed the version that did not work, since now I cannot reproduce the problem; I ran two gcc5 builds, first without CLT and second with CLT, and both seem functional.

I guess either something went wrong between Xcode 8.0 and Xcode 8.1, or removing Xcode was not enough to simulate the case when CLT were not installed; unfortunately I cannot tell what was wrong. :-(

so we'll leave it for now, and if I encounter the problem again I'll preserve the install folder, for further analysis.

**Comment by ilg-ul** [comment]
> You get a warning:

```
Warning: Building gcc from source:
  The bottle needs the Xcode CLT to be installed.
```

I checked and this warning was not issued. Something may still be wrong with the builds, but the stars are in a favourable position, and the compilers are functional.

For reference, the build consoles are packed in the attached archive.

[gcc5-builds.zip](https://github.com/Homebrew/homebrew-versions/files/593429/gcc5-builds.zip)
